asset-transfer-events/application-gateway-typescript/src/app.ts

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright IBM Corp. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as grpc from '@grpc/grpc-js';
+import { ChaincodeEvent, CloseableAsyncIterable, connect, Contract, GatewayError, Identity, Network, Signer, signers } from '@hyperledger/fabric-gateway';
+import * as crypto from 'crypto';
+import { promises as fs } from 'fs';
+import * as path from 'path';
+import { TextDecoder } from 'util';
+
+const channelName = 'mychannel';
+const chaincodeName = 'events';
+const mspId = 'Org1MSP';
+
+// Path to crypto materials.
+const cryptoPath = path.resolve(__dirname, '..', '..', '..', 'test-network', 'organizations', 'peerOrganizations', 'org1.example.com');
+
+// Path to user private key directory.
+const keyDirectoryPath = path.resolve(cryptoPath, 'users', 'User1@org1.example.com', 'msp', 'keystore');
+
+// Path to user certificate.
+const certPath = path.resolve(cryptoPath, 'users', 'User1@org1.example.com', 'msp', 'signcerts', 'cert.pem');
+
+// Path to peer tls certificate.
+const tlsCertPath = path.resolve(cryptoPath, 'peers', 'peer0.org1.example.com', 'tls', 'ca.crt');
+
+// Gateway peer endpoint.
+const peerEndpoint = 'localhost:7051';
+
+const utf8Decoder = new TextDecoder();
+const now = Date.now();
+const assetId = `asset${now}`;
+
+async function main(): Promise<void> {
+    // The gRPC client connection should be shared by all Gateway connections to this endpoint.
+    const client = await newGrpcConnection();
+
+    const gateway = connect({
+        client,
+        identity: await newIdentity(),
+        signer: await newSigner(),
+    });
+    let events: CloseableAsyncIterable<ChaincodeEvent> | undefined;
+
+    try {
+
+        // Get a network instance representing the channel where the smart contract is deployed.
+        const network = gateway.getNetwork(channelName);
+
+        // Get the smart contract from the network.
+        const contract = network.getContract(chaincodeName);
+
+        //Start Listening to events.
+        events = await startEventListening(network)
+
+        // Create a new asset on the ledger.
+        await createAsset(contract);
+
+        // Update an existing asset asynchronously.
+        await transferAssetAsync(contract);
+
+        // Delete asset by assetID.
+        await deleteAssetByID(contract);
+
+    } finally {
+        gateway.close();
+        client.close();
+        events?.close()
```

Comment:
`events` should close before `gateway` (and have a `;` at the end of the line). The eventing is carried out within the scope of a given gateway, so it makes sense to close resources before closing their containing resource.

`client` is the underlying gRPC connection that `gateway` (and `events`) use. It should be closed last after everything that's using it is closed.